### PR TITLE
generate proxy command options

### PR DESCRIPTION
**Pull request by ElizabethOkerio**
This PR contains a list of options the <code> odata-cli </code> generate command will have.

### CODE REVIEW

File: src/OData.Cli/Program.cs [hunk 1]
```diff
@@ -1,18 +1,77 @@
...
+                new Command("generate", "Command to generate proxy classes for OData endpoints.")
+                {
+                    new Option<string>(
+                        new[] { "--metadata", "-m" },
+                        "The URI of the metadata document. The value must be set to a valid service document URI or a local file path."
+                    ){IsRequired = true},
+                    new Option<string>(
+                        new[] { "--custom-headers|--h"},
+                        "Headers that will get sent along with the request when fetching the metadata document from the service."
+                    ),
```
Comment (habbes) [inline src/OData.Cli/Program.cs:22]: What format will the headers be provided in?

File: src/OData.Cli/Program.cs [hunk 2]
```diff
@@ -1,18 +1,77 @@
...
+                    new Option<string>(
+                        new[] { "--proxy|--p"},
+                        "Proxy settings. Format: domain\\user:password@SERVER:PORT."
+                    ),
+                    new Option<string>(
+                        new[] { "--namespace|--ns"},
+                        "The namespace of the client code generated."
```
Comment (habbes) [inline src/OData.Cli/Program.cs:29]: Maybe provide an example here as well?

File: src/OData.Cli/Program.cs [hunk 3]
```diff
@@ -1,18 +1,77 @@
...
+                        new[] { "--no-tracking|-n"},
+                        "The use of this disables entity and property tracking."
+                    ),
+                    new Option<bool?>(
+                        new[] { "--no-alias|-na"},
+                        "This flag indicates whether to disable naming alias."
+                    ),
+                    new Option<bool?>(
+                        new[] { "--internal|-i"},
+                        "If set to true, generated types will have an \"internal\" class modifier instead of \"public\" thereby making them invisible outside the assembly."
```
Comment (habbes) [inline src/OData.Cli/Program.cs:41]: I assume the user won't set this to true per se, but rather they would just provide the option. So maybe it could be reworded? e.g. `Apply the "internal" class modifier on generated classes instead of "public" thereby making them invisible outside the assembly.`

File: src/OData.Cli/Program.cs [hunk 4]
```diff
@@ -1,18 +1,77 @@
...
+                        new[] { "--namespace|--ns"},
+                        "The namespace of the client code generated."
+                    ),
+                    new Option<bool?>(
+                        new[] { "--no-tracking|-n"},
+                        "The use of this disables entity and property tracking."
+                    ),
+                    new Option<bool?>(
+                        new[] { "--no-alias|-na"},
+                        "This flag indicates whether to disable naming alias."
```
Comment (habbes) [inline src/OData.Cli/Program.cs:37]: It's not clear what this option means. Could you provide an example?

File: src/OData.Cli/Program.cs [hunk 5]
```diff
@@ -1,18 +1,77 @@
...
+                        new[] { "--proxy|--p"},
+                        "Proxy settings. Format: domain\\user:password@SERVER:PORT."
+                    ),
+                    new Option<string>(
+                        new[] { "--namespace|--ns"},
+                        "The namespace of the client code generated."
+                    ),
+                    new Option<bool?>(
+                        new[] { "--no-tracking|-n"},
+                        "The use of this disables entity and property tracking."
```
Comment (habbes) [inline src/OData.Cli/Program.cs:33]:
```suggestion
                        "Disable entity and property tracking."
```
I think you should add more info to explain what the impact of disabling entity/property tracking is.

File: src/OData.Cli/Program.cs [hunk 6]
```diff
@@ -1,18 +1,77 @@
...
+                        new[] { "--no-alias|-na"},
+                        "This flag indicates whether to disable naming alias."
+                    ),
+                    new Option<bool?>(
+                        new[] { "--internal|-i"},
+                        "If set to true, generated types will have an \"internal\" class modifier instead of \"public\" thereby making them invisible outside the assembly."
+                    ),
+                    new Option<bool?>(
+                        new[] { "--multiple-files"},
+                        "This files indicates whether to generate the files into multiple files or single."
```
Comment (habbes) [inline src/OData.Cli/Program.cs:45]:
```suggestion
                        "Split the generated classes into separate files instead of generating all the code in a single file."
```

Comment (habbes) [reply src/OData.Cli/Program.cs:45]: Maybe you should also have a short-form alias for this option?

File: src/OData.Cli/Program.cs [hunk 7]
```diff
@@ -1,18 +1,77 @@
...
+                    ),
+                    new Option<string>(
+                        new[] { "-eoi"},
+                        "Comma-separated list of the names of operation imports to exclude from the generated code."
```
Comment (habbes) [inline src/OData.Cli/Program.cs:49]: I think you should have a full-form alias for this option, e.g. `--exclude-operation-imports`. And maybe also provide an example in the description.

File: src/OData.Cli/Program.cs [hunk 8]
```diff
@@ -1,18 +1,77 @@
...
+                    ),
+                    new Option<string>(
+                        new[] { "-ebo"},
+                        "Comma-separated list of the names of bound operations to exclude from the generated code."
```
Comment (habbes) [inline src/OData.Cli/Program.cs:53]: Consider having a long-form option name and an example here as well.

File: src/OData.Cli/Program.cs [hunk 9]
```diff
@@ -1,18 +1,77 @@
...
+                    ),
+                    new Option<string>(
+                        new[] { "-est"},
+                        "Comma-separated list of the names of entity types to exclude from the generated code."
```
Comment (habbes) [inline src/OData.Cli/Program.cs:57]: Consider having a long-form option name and an example here as well.

File: src/OData.Cli/Program.cs [hunk 10]
```diff
@@ -1,18 +1,77 @@
...
+                    ),
+                    new Option<string>(
+                        new[] { "-efi"},
+                        "Comma-separated list of the names of function imports to exclude from the generated code."
```
Comment (habbes) [inline src/OData.Cli/Program.cs:61]: Consider having a long-form option name and an example here as well.

File: src/OData.Cli/Program.cs [hunk 11]
```diff
@@ -1,18 +1,77 @@
...
+                    ),
+                    new Option<string>(
+                        new[] { "-eai"},
+                        "Comma-separated list of the names of action imports to exclude from the generated code."
```
Comment (habbes) [inline src/OData.Cli/Program.cs:65]: Consider having a long-form option name and an example here as well.

File: src/OData.Cli/Program.cs [hunk 12]
```diff
@@ -1,18 +1,77 @@
...
+                    ),
+                    new Option<string>(
+                        new[] { "-iue"},
+                        "This flag indicates whether to ignore unexpected elements and attributes in the metadata document and generate the client code if any."
```
Comment (habbes) [inline src/OData.Cli/Program.cs:69]: Consider having a long-form option name and an example here as well.

File: src/OData.Cli/Program.cs [hunk 13]
```diff
@@ -1,18 +1,77 @@
...
+                    )
+                }
+            };
 
-            app.Name = "odata-cli";
-            app.Description = "Client Proxy generator for OData endpoints.";
-
-            // Set the arguments to display the description and help text
-            app.HelpOption("-?|-h|--help");
```
Comment (habbes) [inline src/OData.Cli/Program.cs:15]: Why did you remove these?

File: src/OData.Cli/OData.Cli.csproj [hunk 14]
```diff
@@ -8,7 +8,7 @@
   </PropertyGroup>
 
   <ItemGroup>
-    <PackageReference Include="Microsoft.Extensions.CommandLineUtils" Version="1.1.1" />
+    <PackageReference Include="System.CommandLine" Version="2.0.0-beta1.21308.1" />
```
Comment (gathogojr) [inline src/OData.Cli/OData.Cli.csproj:11]: Should we be referencing a beta version?

File: src/OData.Cli/Program.cs [hunk 15]
```diff
@@ -1,18 +1,77 @@
...
+            var app = new RootCommand
+            {
+                new Command("generate", "Command to generate proxy classes for OData endpoints.")
+                {
+                    new Option<string>(
+                        new[] { "--metadata", "-m" },
+                        "The URI of the metadata document. The value must be set to a valid service document URI or a local file path."
+                    ){IsRequired = true},
+                    new Option<string>(
+                        new[] { "--custom-headers|--h"},
```
Comment (gathogojr) [inline src/OData.Cli/Program.cs:20]: How come the shortened option starts with `-` in some cases and `--` in some other cases? E.g. `-m` vs `--h`

File: src/OData.Cli/Program.cs (hunk 9 above)
Comment (gathogojr) [reply src/OData.Cli/Program.cs:57]: `excluded-entity-types` feels like a good choice

File: src/OData.Cli/Program.cs (hunk 12 above)
Comment (gathogojr) [reply src/OData.Cli/Program.cs:69]: `ignore-unexpected-elements` feels like a good choice

**Comment by gathogojr** [comment]
@ElizabethOkerio A few points to consider:
1. How does someone specify service name?
2. Should we not allow someone the option of specifying a settings config file to be used?
3. Should we not allow someone the option of specifying a file location for the metadata file, i.e., not just an web address?
4. Should we not allow someone the option of saving the settings to a config file?
5. Should we not allow someone the option of specifying the name of the proxy classes file?
6. Should we not allow someone the option of dictating whether C# casing style should be applied?
7. Should we not allow someone the option of emitting T4 files? For example, if they choose to use Visual Studio after

**Comment by ElizabethOkerio** [comment]
> @ElizabethOkerio A few points to consider:
> 
> 1. How does someone specify service name?
> 2. Should we not allow someone the option of specifying a settings config file to be used?
> 3. Should we not allow someone the option of specifying a file location for the metadata file, i.e., not just a web address?
> 4. Should we not allow someone the option of saving the settings to a config file?
> 5. Should we not allow someone the option of specifying the name of the proxy classes file?
> 6. Should we not allow someone the option of dictating whether C# casing style should be applied?
> 7. Should we not allow someone the option of emitting T4 files? For example, if they choose to use Visual Studio after

@gathogojr  I don't think we'll need a service name - this was mostly for the connected service. I could be wrong though. For the settings config file also - Am not sure we need it. This tool will be generating only the proxy classes and the metadata file - I don't think it will have the ConnectedService.json file. I didn't see the need for the settings file.  With the metadata command, you can provide a location to a file or a web address.

File: src/OData.Cli/OData.Cli.csproj (hunk 14 above)
Comment (ElizabethOkerio) [reply src/OData.Cli/OData.Cli.csproj:11]: The latest release for this package is in beta. The previous release was alpha.

File: src/OData.Cli/Program.cs (hunk 3 above)
Comment (ElizabethOkerio) [reply src/OData.Cli/Program.cs:40]: have set it as a bool - they should pass true or false

**Comment by gathogojr** [comment]
> > @ElizabethOkerio A few points to consider:
> > 
> > 1. How does someone specify service name?
> > 2. Should we not allow someone the option of specifying a settings config file to be used?
> > 3. Should we not allow someone the option of specifying a file location for the metadata file, i.e., not just a web address?
> > 4. Should we not allow someone the option of saving the settings to a config file?
> > 5. Should we not allow someone the option of specifying the name of the proxy classes file?
> > 6. Should we not allow someone the option of dictating whether C# casing style should be applied?
> > 7. Should we not allow someone the option of emitting T4 files? For example, if they choose to use Visual Studio after
> 
> @gathogojr I don't think we'll need a service name - this was mostly for the connected service. I could be wrong though. For the settings config file also - Am not sure we need it. This tool will be generating only the proxy classes and the metadata file - I don't think it will have the ConnectedService.json file. I didn't see the need for the settings file. With the metadata command, you can provide a location to a file or a web address.

@ElizabethOkerio Thank you for you response. You make valid points. The thing I have a different view about is the config file. Especially because of things like excluded schema types, operation imports, functions and actions that would require the user to do a lot of types, the config file would come in handy since it'd make it many times easier for the user to use the tools subsequent times

File: src/OData.Cli/Program.cs [hunk 16]
```diff
@@ -1,18 +1,93 @@
...
+        static async Task Main(string[] args)
         {
-            var app = new CommandLineApplication();
-
-            app.Name = "odata-cli";
-            app.Description = "Client Proxy generator for OData endpoints.";
-
-            // Set the arguments to display the description and help text
-            app.HelpOption("-?|-h|--help");
+            System.Console.WriteLine("popos");
```
Comment (KenitoInc) [inline src/OData.Cli/Program.cs:10]: Why do we have this line?

File: src/OData.Cli/Program.cs [hunk 17]
```diff
@@ -1,18 +1,93 @@
...
+            var app = new RootCommand
```
Comment (KenitoInc) [inline src/OData.Cli/Program.cs:11]: We should have a `--help` option

File: src/OData.Cli/Program.cs [hunk 18]
```diff
@@ -1,18 +1,91 @@
...
+                    new Option<string>(
+                        new[] { "--file-name|-fn"},
+                        "The file name of the generated proxy classes."
+                    ),
+                    new Option<bool>(
+                        new[] { "--c#-casing-style|-ccs"},
+                        "This flag indicates whether to use the C# casing style or not."
+                    ),
+                    new Option<bool>(
+                        new[] { "--include-t4-files|-itt"},
```
Comment (gathogojr) [inline src/OData.Cli/Program.cs:83]: @ElizabethOkerio Can you get the opinion of others whether we need to provide the option of emitting T4 files when someone is using this tool? Alternatively we could leave out the option and add it later if it's requested

File: src/OData.Cli/Program.cs [hunk 19]
```diff
@@ -1,18 +1,91 @@
...
+                        new[] { "--namespace|-ns"},
+                        "The namespace of the client code generated. Example:ODataCliCodeGeneratorSample.NorthWindModel or ODataCliCodeGeneratorSample or it could be a name related to the OData endpoint."
+                    ),
+                    new Option<bool?>(
+                        new[] { "--no-tracking|-n"},
+                        "Disables entity and property tracking. Entity tracking enables the client to track an entity in DataServiceContext while property tracking enables the client to track only top-level properties."
+                    ),
+                    new Option<bool?>(
+                        new[] { "--no-alias|-na"},
+                        "Disables naming alias."
```
Comment (habbes) [inline src/OData.Cli/Program.cs:40]: It's not clear what this option means. What is "naming alias"?

Comment (habbes) [reply src/OData.Cli/Program.cs:40]: I think the option is used to enable PascalCase (i.e. `alertType` would be renamed to `AlertType` when this option is enabled).

File: src/OData.Cli/Program.cs [hunk 20]
```diff
@@ -1,18 +1,91 @@
...
+                    new Option<bool>(
+                        new[] { "--ignore-unexpected-elements|-iue"},
+                        "This flag indicates whether to ignore unexpected elements and attributes in the metadata document and generate the client code if any."
+                    ),
+                    new Option<string>(
+                        new[] { "--file-name|-fn"},
+                        "The file name of the generated proxy classes."
+                    ),
+                    new Option<bool>(
+                        new[] { "--c#-casing-style|-ccs"},
```
Comment (habbes) [inline src/OData.Cli/Program.cs:79]: I think `csharp-casing` or `cs-casing` instead of `c#` to avoid the special character? Not sure it's a big deal though.

File: src/OData.Cli/Program.cs (hunk 19 above)
Comment (habbes) [reply src/OData.Cli/Program.cs:40]: I see there's another option for c#/PascalCasing. So I'm still unsure what this option does.

**Review by habbes — APPROVED** [review]
Minor comments. But looks okay.

File: src/OData.Cli/Program.cs (hunk 4 above)
Comment (ElizabethOkerio) [reply src/OData.Cli/Program.cs]: disables upper camel casing. it is enabled by default.

File: src/OData.Cli/Program.cs [hunk 21]
```diff
@@ -1,18 +1,87 @@
...
+            {
+                new Command("generate", "Command to generate proxy classes for OData endpoints.")
+                {
+                    new Option<string>(
+                        new[] { "--metadata|-m"},
+                        "The URI of the metadata document. The value must be set to a valid service document URI or a local file path."
+                    ){IsRequired = true},
+                    new Option<string>(
+                        new[] { "--load-settings-from-file|-lsff" },
+                        "The URI to the settings file."
```
Comment (gathogojr) [inline src/OData.Cli/Program.cs:20]:
```suggestion
                        "The location of the configuration settings file."
```
Because URI suggests a web location

File: src/OData.Cli/Program.cs [hunk 22]
```diff
@@ -1,18 +1,87 @@
...
+                    ),
+                    new Option<string>(
+                        new[] { "--excluded-function-imports|-efi"},
+                        "Comma-separated list of the names of function imports to exclude from the generated code.Example: FunctionImport1,FunctionImport2"
+                    ),
+                    new Option<string>(
+                        new[] { "--excluded-action-imports|-eai"},
+                        "Comma-separated list of the names of action imports to exclude from the generated code.Example: ActionImport1,ActionImport2"
+                    ),
+                    new Option<bool>(
```
Comment (gathogojr) [inline src/OData.Cli/Program.cs:70]: Should this be `new Option<bool?>(` as well?

File: src/OData.Cli/Program.cs [hunk 23]
```diff
@@ -1,18 +1,87 @@
...
+                    new Option<string>(
+                        new[] { "--file-name|-fn"},
+                        "The file name of the generated proxy classes."
+                    ),
+                    new Option<string>(
+                        new[] { "--outputdir|-o"},
+                        "Full path to output directory."
+                    )
+                }
+            };
```
Comment (gathogojr) [inline src/OData.Cli/Program.cs:83]:
```suggestion
            };

```

**Review by gathogojr — APPROVED** [review]
Left some comments otherwise looks good
